File: packages/fast_lifecycle/android/src/main/kotlin/com/flutterbazaar/fast_lifecycle/FastLifecyclePlugin.kt
package com.flutterbazaar.fast_lifecycle

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Android 系统原生层入口（架构第 ④ 层）。
 *
 * ## 通道规范（需求文档 §3）
 * | 通道 | 名称 | 职责 |
 * |------|------|------|
 * | EventChannel | fast_lifecycle/events | 流式推送 rawState（唯一事件通道） |
 * | MethodChannel | fast_lifecycle/control | startListening / stopListening |
 *
 * ## 架构红线
 * - 禁止用 MethodChannel 推送持续生命周期事件
 * - 禁止翻译 / 重命名 rawState（输出 AppLifecycleState 原字符串）
 * - stopListening / Engine 分离时必须移除原生订阅
 */
class FastLifecyclePlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware {

    private lateinit var applicationContext: Context
    private lateinit var controlChannel: MethodChannel
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null

    private var tracker: AppLifecycleTracker? = null
    private var pendingActivity: Activity? = null
    private var listeningRequested = false
    private var listenWindowId: String? = null
    private var listenIsMainWindow: Boolean = true

    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        controlChannel = MethodChannel(binding.binaryMessenger, CONTROL_CHANNEL)
        controlChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel?.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            METHOD_START_LISTENING -> {
                parseListenArguments(call.arguments)
                requestNativeListening()
                result.success(null)
            }
            METHOD_STOP_LISTENING -> {
                stopNativeListening()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        parseListenArguments(arguments)
        requestNativeListening()
    }

    override fun onCancel(arguments: Any?) {
        stopNativeListening()
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        stopNativeListening()
        controlChannel.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        eventSink = null
    }

    // ── ActivityAware：绑定 Flutter Activity 生命周期 ─────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        pendingActivity = binding.activity
        tracker?.bindActivity(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        tracker?.unbindActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        pendingActivity = binding.activity
        tracker?.bindActivity(binding.activity)
    }

    override fun onDetachedFromActivity() {
        pendingActivity = null
        tracker?.unbindActivity()
    }

    // ── 内部方法 ──────────────────────────────────────────────────────

    private fun parseListenArguments(arguments: Any?) {
        if (arguments !is Map<*, *>) return
        listenWindowId = arguments["windowId"] as? String
        listenIsMainWindow = arguments["isMainWindow"] as? Boolean ?: true
    }

    /**
     * 标记需要监听；仅当 [EventChannel] 已就绪（[eventSink] 非空）时才启动 Tracker，
     * 避免 MethodChannel 先于 EventChannel 订阅时丢失事件。
     */
    private fun requestNativeListening() {
        listeningRequested = true
        tryStartNativeListening()
    }

    private fun tryStartNativeListening() {
        if (tracker != null || !listeningRequested || eventSink == null) {
            return
        }

        val app = applicationContext.applicationContext as Application
        tracker = AppLifecycleTracker(
            onEvent = ::emitEvent,
            windowId = listenWindowId,
            isMainWindow = listenIsMainWindow,
        )
        tracker?.start(app)
        pendingActivity?.let { tracker?.bindActivity(it) }
    }

    private fun emitEvent(payload: Map<String, Any?>) {
        val sink = eventSink ?: return
        if (Looper.myLooper() == Looper.getMainLooper()) {
            sink.success(payload)
            return
        }
        mainHandler.post {
            eventSink?.success(payload)
        }
    }

    private fun stopNativeListening() {
        listeningRequested = false
        val app = applicationContext.applicationContext as? Application ?: return
        tracker?.stop(app)
        tracker?.unbindActivity()
        tracker = null
    }

    companion object {
        private const val EVENT_CHANNEL = "fast_lifecycle/events"
        private const val CONTROL_CHANNEL = "fast_lifecycle/control"
        private const val METHOD_START_LISTENING = "startListening"
        private const val METHOD_STOP_LISTENING = "stopListening"
    }
}

File: packages/fast_lifecycle/ios/fast_lifecycle/Sources/fast_lifecycle/FastLifecyclePlugin.swift
import Flutter
import UIKit

// MARK: - 跨端载荷构造（字段名与 Dart NativePayloadKeys 一致）

/// EventChannel 标准载荷构造器。
///
/// - Important: [rawState] 必须为 UIApplicationDelegate 方法名原字符串，禁止翻译。
private enum LifecycleEventPayload {
  static func build(
    rawState: String,
    source: String = "ios",
    lifecycleScope: String = "application",
    windowId: String? = nil,
    isMainWindow: Bool = true
  ) -> [String: Any] {
    var extra: [String: Any] = [
      "lifecycleScope": lifecycleScope,
      "isMainWindow": isMainWindow,
    ]
    if let windowId {
      extra["windowId"] = windowId
    }
    return [
      "source": source,
      "rawState": rawState,
      "extra": extra,
    ]
  }
}

// MARK: - UIApplication 生命周期采集

/// iOS 原生生命周期采集器。
///
/// rawState 一一对应 [UIApplicationDelegate] 方法名（需求文档 §4.1）：
/// - applicationDidFinishLaunching
/// - applicationDidBecomeActive
/// - applicationWillResignActive
/// - applicationDidEnterBackground
/// - applicationWillEnterForeground
/// - applicationWillTerminate
private final class AppLifecycleTracker {
  private var observers: [NSObjectProtocol] = []
  private var lastRawState: String?
  private let onEvent: ([String: Any]) -> Void
  private let windowId: String?
  private let isMainWindow: Bool

  init(
    onEvent: @escaping ([String: Any]) -> Void,
    windowId: String?,
    isMainWindow: Bool
  ) {
    self.onEvent = onEvent
    self.windowId = windowId
    self.isMainWindow = isMainWindow
  }

  func start() {
    guard observers.isEmpty else { return }

    let center = NotificationCenter.default
    let mappings: [(Notification.Name, String)] = [
      (UIApplication.didFinishLaunchingNotification, "applicationDidFinishLaunching"),
      (UIApplication.didBecomeActiveNotification, "applicationDidBecomeActive"),
      (UIApplication.willResignActiveNotification, "applicationWillResignActive"),
      (UIApplication.didEnterBackgroundNotification, "applicationDidEnterBackground"),
      (UIApplication.willEnterForegroundNotification, "applicationWillEnterForeground"),
      (UIApplication.willTerminateNotification, "applicationWillTerminate"),
    ]

    for (name, rawState) in mappings {
      let token = center.addObserver(
        forName: name,
        object: nil,
        queue: .main
      ) { [weak self] _ in
        self?.emit(rawState: rawState)
      }
      observers.append(token)
    }

    emitCurrentStateIfNeeded()
  }

  func stop() {
    let center = NotificationCenter.default
    for token in observers {
      center.removeObserver(token)
    }
    observers.removeAll()
    lastRawState = nil
  }

  private func emitCurrentStateIfNeeded() {
    let app = UIApplication.shared
    switch app.applicationState {
    case .active:
      emit(rawState: "applicationDidBecomeActive")
    case .inactive:
      emit(rawState: "applicationWillResignActive")
    case .background:
      emit(rawState: "applicationDidEnterBackground")
    @unknown default:
      break
    }
  }

  private func emit(rawState: String) {
    if lastRawState == rawState { return }
    lastRawState = rawState
    onEvent(
      LifecycleEventPayload.build(
        rawState: rawState,
        windowId: windowId,
        isMainWindow: isMainWindow
      )
    )
  }
}

// MARK: - Flutter Plugin（系统原生层入口）

/// iOS 系统原生层（架构第 ④ 层）。
///
/// - EventChannel `fast_lifecycle/events`：唯一事件推送通道
/// - MethodChannel `fast_lifecycle/control`：仅 startListening / stopListening
public class FastLifecyclePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  private var tracker: AppLifecycleTracker?
  private var listenWindowId: String?
  private var listenIsMainWindow = true

  public static func register(with registrar: FlutterPluginRegistrar) {
    let controlChannel = FlutterMethodChannel(
      name: "fast_lifecycle/control",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "fast_lifecycle/events",
      binaryMessenger: registrar.messenger()
    )

    let instance = FastLifecyclePlugin()
    registrar.addMethodCallDelegate(instance, channel: controlChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "startListening":
      parseListenArguments(call.arguments)
      startNativeListening()
      result(nil)
    case "stopListening":
      stopNativeListening()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    eventSink = events
    parseListenArguments(arguments)
    startNativeListening()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    stopNativeListening()
    eventSink = nil
    return nil
  }

  private func parseListenArguments(_ arguments: Any?) {
    guard let map = arguments as? [String: Any] else { return }
    listenWindowId = map["windowId"] as? String
    listenIsMainWindow = map["isMainWindow"] as? Bool ?? true
  }

  private func startNativeListening() {
    guard tracker == nil else { return }

    tracker = AppLifecycleTracker(
      onEvent: { [weak self] payload in
        self?.eventSink?(payload)
      },
      windowId: listenWindowId,
      isMainWindow: listenIsMainWindow
    )
    tracker?.start()
  }

  private func stopNativeListening() {
    tracker?.stop()
    tracker = nil
  }
}
